File: AlgorithmsOnGraphs-UCSD/week4/dijkstra.py
 #Uses python3
import sys

def minDistance(visited, distance):
    min_vertex = len(distance)
    min_distance = sys.maxsize

    for v in range(len(distance)):
        if not visited[v] and distance[v] < min_distance:
            min_distance = distance[v]
            min_vertex = v    
    return min_vertex


def distance(adj, cost, s, t):
    #write your code here
    '''
        #assuming directed graph
        adj --> 0 1 2 3 
                2 4 6 5   
                (0 is index of list which represents src index)
                (2 is destination index)
       cost --> 0 1 2
                5 3 7
                (0 is src index which represents weight)
    '''
    number_of_vertices = len(adj)
    distance = [sys.maxsize] * number_of_vertices
    visited = [False] * number_of_vertices
    distance[s] = 0 #src distance
    for _ in range(number_of_vertices - 1):
        u = minDistance(visited, distance) 
        if u == number_of_vertices:
            break
        visited[u] = True
        i = 0 # for all neighbours from u. Check cost[u][i]
        for v in adj[u]:
            if not visited[v] and distance[v] > distance[u] + cost[u][i]:
                distance[v] = distance[u] + cost[u][i]
            i += 1
    return distance[t] if distance[t] != sys.maxsize else -1

 
if __name__ == '__main__':
    user_input = sys.stdin.read()
    data = list(map(int, user_input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(
        zip(zip(data[0:(3 * m):3], data[1:(3 * m):3]), data[2:(3 * m):3]))
    data = data[3 * m:]
    adj = [[] for _ in range(n)]
    cost = [[] for _ in range(n)]
    for ((a, b), w) in edges:
        adj[a - 1].append(b - 1)
        cost[a - 1].append(w)
    s, t = data[0] - 1, data[1] - 1
    print(distance(adj, cost, s, t))

File: AlgorithmsOnGraphs-UCSD/week4/dijkstra.cpp
#include <iostream>
#include <vector>
#include <queue>

using std::vector;
using std::queue;
using std::pair;
using std::priority_queue;

int minDistance(const vector<int> &distances, const vector<bool> &visited){
    int min_dist = INT32_MAX;
    int min_index = 0;

    for(int i = 0; i < distances.size(); i++){
        if(!visited[i] && distances[i] < min_dist){
            min_dist = distances[i];
            min_index = i;
        }
    }
    return min_index;
}

int distance(vector<vector<int> > &adj, vector<vector<int> > &cost, int s, int t) {
    //write your code here
    vector<int> distances(adj.size(), INT32_MAX);
    vector<bool> visited(adj.size(), false);
    distances[s] = 0;

    for(int i = 0; i < adj.size() - 1; i++){
        int u = minDistance(distances, visited);
        if(u == adj.size())
            break;  
        int j = 0;
        visited[u] = true;
        for(auto it = adj[u].begin(); it != adj[u].end(); it++){
            if(!visited[*it] && distances[*it] > distances[u] + cost[u][j]){
                distances[*it] = distances[u] + cost[u][j];
            }
            j++;
        }
    }

    if(distances[t] != INT32_MAX)
        return distances[t];
    else return -1;
}

int main() {
  int n, m;
  std::cin >> n >> m;
  vector<vector<int> > adj(n, vector<int>());
  vector<vector<int> > cost(n, vector<int>());
  for (int i = 0; i < m; i++) {
    int x, y, w;
    std::cin >> x >> y >> w;
    adj[x - 1].push_back(y - 1);
    cost[x - 1].push_back(w);
  }
  int s, t;
  std::cin >> s >> t;
  s--, t--;
  std::cout << distance(adj, cost, s, t);
}
